packages/roosterjs-editor-plugins/lib/CustomReplace/CustomReplace.ts

```diff
@@ -0,0 +1,157 @@
+import { Editor, EditorPlugin } from 'roosterjs-editor-core';
+import { PluginEvent, PluginEventType, PluginKeyPressEvent } from 'roosterjs-editor-types';
+
+type Replacement = {
+    sourceString: string,
+    replacementHTML: string,
+    matchSourceCaseSensitive: boolean
+}
+
+const defaultReplacements: Replacement[] = [
+    {
+        sourceString: ':)',
+        replacementHTML: '🙂',
+        matchSourceCaseSensitive: true,
+    },
+    {
+        sourceString: ';)',
+        replacementHTML: '😉',
+        matchSourceCaseSensitive: true,
+    },
+    {
+        sourceString: ':O',
+        replacementHTML: '😲',
+        matchSourceCaseSensitive: true,
+    },
+    {
+        sourceString: ':o',
+        replacementHTML: '😯',
+        matchSourceCaseSensitive: true,
+    },
+    {
+        sourceString: '<3',
+        replacementHTML: '❤️',
+        matchSourceCaseSensitive: true,
+    },
+]
+
+/**
+ * Wrapper for CustomReplaceContentEditFeature that provides an API for updating the
+ * content edit feature
+ */
+export default class CustomReplacePlugin implements EditorPlugin {
+    private longestReplacementLength: number;
+    private editor: Editor;
+
+    /**
+     * Create instance of CustomReplace plugin
+     * @param features An optional feature set to determine which features the plugin should provide
+     */
+    constructor(private replacements: Replacement[] = defaultReplacements) {
+        this.longestReplacementLength = getLongestReplacementSourceLength(this.replacements);
+    }
+
+    updateReplacements(newReplacements: Replacement[]) {
+        this.replacements = newReplacements;
+        this.longestReplacementLength = getLongestReplacementSourceLength(this.replacements);
+    }
+
+    /**
+     * Get a friendly name of  this plugin
+     */
+    getName() {
+        return 'CustomReplace';
+    }
+
+    /**
+     * Initialize this plugin
+     * @param editor The editor instance
+     */
+    public initialize(editor: Editor): void {
+        this.editor = editor;
+    }
+
+    /**
+     * Dispose this plugin
+     */
+    public dispose(): void {
+    }
+
+    public onPluginEvent(event: PluginEvent) {
+        if (this.editor.isInIME() || event.eventType != PluginEventType.KeyPress) {
+            return;
+        }
+
+        // This handler fires before the event's native behaviour -- defer until after it completes
+        setTimeout(() => this.handlePluginEvent(event), 0)
+    }
+
+    private handlePluginEvent(event: PluginKeyPressEvent) {
+        // Get the matching replacement
+        const range = document.getSelection().getRangeAt(0);
+        if (range == null) return;
+        const searcher = this.editor.getContentSearcherOfCursor();
+        const stringToSearch = searcher.getSubStringBefore(this.longestReplacementLength);
```

Comment:
In very early implementation, it work like this. But then I found doing a string search here maybe a little low-perf. So I changed it by only search text in the nearest text node. It doesn't work in some corner case (e.g. text is splitted into several text nodes), but it works in most case with high efficency.